File: code.py
from flask import Flask, request, jsonify
import requests
from urllib.parse import urlparse

app = Flask(__name__)

# Add this function to handle OPTIONS requests
@app.before_request
def handle_preflight():
    if request.method == "OPTIONS":
        res = app.make_response("")
        res.headers['Access-Control-Allow-Origin'] = '*'
        res.headers['Access-Control-Allow-Methods'] = 'GET, OPTIONS'
        res.headers['Access-Control-Allow-Headers'] = 'Content-Type'
        return res

@app.route('/code.py', methods=['GET', 'OPTIONS']) # Add 'OPTIONS' here
def fetch_html():
    url = request.args.get('url', '')
    
    # Validate URL
    if not url:
        return jsonify({
            'html': '',
            'size': '0 B',
            'error': 'No URL provided.'
        })
    
    parsed = urlparse(url)
    if not parsed.scheme or not parsed.netloc:
        return jsonify({
            'html': '',
            'size': '0 B',
            'error': 'Invalid URL format. Please use http:// or https://'
        })
    
    try:
        # Fetch the content
        response = requests.get(url, timeout=30, headers={
            'User-Agent': 'Mozilla/5.0 (compatible; SourceViewer/1.0)'
        })
        response.raise_for_status()
        
        html_content = response.text
        
        res = jsonify({
            'html': html_content,
            'size': f"{len(html_content)} B",
            'error': None
        })
        # Add CORS headers to the main response
        res.headers['Access-Control-Allow-Origin'] = '*'
        res.headers['Access-Control-Allow-Methods'] = 'GET, OPTIONS'
        res.headers['Access-Control-Allow-Headers'] = 'Content-Type'
        return res
        
    except Exception as e:
        res = jsonify({
            'html': '',
            'size': '0 B',
            'error': f'Failed to fetch content: {str(e)}'
        })
        res.headers['Access-Control-Allow-Origin'] = '*'
        res.headers['Access-Control-Allow-Methods'] = 'GET, OPTIONS'
        res.headers['Access-Control-Allow-Headers'] = 'Content-Type'
        return res

if __name__ == '__main__':
    app.run(host='0.0.0.0', port=10000)

File: code.php
<?php
// code.php
// A simple, CORS-enabled PHP proxy to fetch and return the HTML source of any URL.

// Enable CORS so your frontend (e.g., on GitHub Pages) can access this script.
header('Access-Control-Allow-Origin: *');
header('Access-Control-Allow-Methods: GET, POST');
header('Access-Control-Allow-Headers: Content-Type');
header('Content-Type: application/json');

// Get the target URL from the query string
$url = $_GET['url'] ?? '';

// Initialize response data
$response = [
    'html' => '',
    'size' => '0 B',
    'time' => '0.00',
    'error' => null
];

// Start timing the request
$startTime = microtime(true);

// Validate the URL
if (empty($url)) {
    $response['error'] = 'No URL provided.';
    echo json_encode($response);
    exit;
}

if (!filter_var($url, FILTER_VALIDATE_URL)) {
    $response['error'] = 'Invalid URL format. Please use http:// or https://';
    echo json_encode($response);
    exit;
}

// Attempt to fetch the content using file_get_contents
$htmlContent = '';

$context = stream_context_create([
    'http' => [
        'timeout' => 30,
        'user_agent' => 'Mozilla/5.0 (compatible; SourceViewer/1.0)'
    ]
]);

$htmlContent = @file_get_contents($url, false, $context);

if ($htmlContent === false) {
    $response['error'] = 'Failed to fetch content. The target server may be unreachable or blocking requests.';
} else {
    // Calculate stats
    $endTime = microtime(true);
    $response['time'] = number_format($endTime - $startTime, 2);
    $response['size'] = formatBytes(strlen($htmlContent));
    
    // Escape the HTML so it can be safely sent as JSON and displayed in the browser
    $response['html'] = htmlspecialchars($htmlContent);
}

// Send the JSON response back to the frontend
echo json_encode($response);

// Helper function to format bytes
function formatBytes($bytes, $precision = 2) {
    $units = ['B', 'KB', 'MB', 'GB'];
    $bytes = max($bytes, 0);
    $pow = floor(($bytes ? log($bytes) : 0) / log(1024));
    $pow = min($pow, count($units) - 1);
    $bytes /= pow(1024, $pow);
    return round($bytes, $precision) . ' ' . $units[$pow];
}
?>
